### PR TITLE
Hibernate_Inheritance_HW

### PR DESCRIPTION
**Pull request by AlexPravda1**
- Inheritance annotations added to models
- Dao level implemeneted
- Prepared Query used

### CODE REVIEW

File: src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java [hunk 1]
```diff
@@ -12,11 +15,39 @@ public AnimalDaoImpl(SessionFactory sessionFactory) {
...
     public List<Animal> findByNameFirstLetter(Character character) {
-        return null;
+        try (Session session = sessionFactory.openSession()) {
+            List<Animal> animalsFromDB = session
+                    .createQuery("SELECT DISTINCT a FROM Animal a", Animal.class)
+                    .getResultList();
+
+            return animalsFromDB.stream().filter(n -> n.getName().toLowerCase()
+                    .startsWith(String.valueOf(character).toLowerCase()))
+                    .collect(Collectors.toList());
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java:47]: It's bad practice to get all animals from db and than filter them, let's create a query that will get the necessary records from the database

File: src/main/java/core/basesyntax/dao/figure/FigureDaoImpl.java [hunk 2]
```diff
@@ -12,11 +14,37 @@ public FigureDaoImpl(SessionFactory sessionFactory) {
...
+            }
+        }
     }
 
     @Override
     public List<T> findByColor(String color, Class<T> clazz) {
-        return null;
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("FROM "
+                            + clazz.getName().substring(clazz.getName().lastIndexOf(".") + 1)
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/dao/figure/FigureDaoImpl.java:41]: maybe it's better to use clazz.getName().getSimpleName() here?

**Review by MaliukDaria — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/dao/ma/CoachDaoImpl.java [hunk 3]
```diff
@@ -11,6 +12,14 @@ public CoachDaoImpl(SessionFactory sessionFactory) {
 
     @Override
     public List<Coach> findByExperienceGreaterThan(int years) {
-        return null;
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("SELECT c FROM Coach c"
+                            + " WHERE c.experience > :exp", Coach.class)
+                    .setParameter("exp", years)
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dao/ma/CoachDaoImpl.java:18]:
```suggestion
                    .setParameter("experience", years)
```

File: src/main/java/core/basesyntax/dao/ma/CoachDaoImpl.java [hunk 4]
```diff
@@ -11,6 +12,14 @@ public CoachDaoImpl(SessionFactory sessionFactory) {
...
+                    .getResultList();
+        } catch (Exception e) {
+            throw new RuntimeException("Can't find Coach Person with exp more years than: "
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dao/ma/CoachDaoImpl.java:21]: same - don't use shortenings

File: src/main/java/core/basesyntax/model/zoo/Animal.java [hunk 5]
```diff
@@ -1,9 +1,29 @@
...
 
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+
+@Entity
+@Inheritance (strategy = InheritanceType.TABLE_PER_CLASS)
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/zoo/Animal.java:11]: should we use TABLE_PER_CLASS here?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Pull request closed without merging**